mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassin.java

```diff
@@ -30,32 +31,76 @@
 
 import com.github.fge.lambdas.Throwing;
 
-public class SpamAssassin {
+import reactor.core.Disposable;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.UnicastProcessor;
+import reactor.core.scheduler.Schedulers;
+import reactor.util.function.Tuple2;
+
+public class SpamAssassin implements Closeable {
 
     private final MetricFactory metricFactory;
-    private final SpamAssassinConfiguration spamAssassinConfiguration;
+    private Optional<UnicastProcessor<Tuple2<InputStream, Username>>> hamQueue;
+    private Optional<UnicastProcessor<Tuple2<InputStream, Username>>> spamQueue;
+    private Optional<Disposable> hamHandle;
+    private Optional<Disposable> spamHandle;
 
     @Inject
     public SpamAssassin(MetricFactory metricFactory, SpamAssassinConfiguration spamAssassinConfiguration) {
         this.metricFactory = metricFactory;
-        this.spamAssassinConfiguration = spamAssassinConfiguration;
+
+        createListeners(spamAssassinConfiguration);
     }
 
-    public void learnSpam(List<InputStream> messages, Username username) {
-        if (spamAssassinConfiguration.isEnable()) {
-            Host host = spamAssassinConfiguration.getHost().get();
-            SpamAssassinInvoker invoker = new SpamAssassinInvoker(metricFactory, host.getHostName(), host.getPort());
-            messages
-                .forEach(Throwing.consumer(message -> invoker.learnAsSpam(message, username)));
-        }
+    public void learnSpam(Flux<InputStream> messages, Username username) {
+        spamQueue.ifPresent(queue ->
```

Comment:
Shouldn't we fail if spamQueue is not present?

Reply:
it was not the case, we do not have it if SA is disabled